interfaces/prompting/constraints.go

```diff
@@ -0,0 +1,263 @@
+			return nil, fmt.Errorf("unsupported permission for %q interface: %q", iface, perm)
+		}
+		permsSet[perm] = true
+	}
+	if len(permsSet) == 0 {
+		return nil, ErrPermissionsListEmpty
+	}
+	permissionsList := make([]string, 0, len(permsSet))
+	for _, perm := range availablePerms {
+		if exists := permsSet[perm]; exists {
+			permissionsList = append(permissionsList, perm)
+		}
+	}
+	return permissionsList, nil
+}
+
+// AbstractPermissionsToAppArmorPermissions returns AppArmor permissions
+// corresponding to the given permissions for the given interface.
+func AbstractPermissionsToAppArmorPermissions(iface string, permissions []string) (interface{}, error) {
+	return abstractPermissionsToAppArmorFilePermissions(iface, permissions)
+}
+
+// AbstractPermissionsToAppArmorFilePermissions returns AppArmor file
+// permissions corresponding to the given permissions for the given interface.
+func abstractPermissionsToAppArmorFilePermissions(iface string, permissions []string) (notify.FilePermission, error) {
+	if len(permissions) == 0 {
+		return notify.FilePermission(0), ErrPermissionsListEmpty
+	}
+	filePermsMap, exists := interfaceFilePermissionsMaps[iface]
+	if !exists {
+		// This should never happen, since iface is checked in the calling function
+		return notify.FilePermission(0), fmt.Errorf("internal error: no AppArmor file permissions map defined for interface: %q", iface)
+	}
+	filePerms := notify.FilePermission(0)
+	for _, perm := range permissions {
+		permMask, exists := filePermsMap[perm]
+		if !exists {
+			// Should not occur, since stored permissions list should have been validated
+			return notify.FilePermission(0), fmt.Errorf("no AppArmor file permission mapping for %q interface with abstract permission: %q", iface, perm)
+		}
+		filePerms |= permMask
+	}
+	if filePerms&(notify.AA_MAY_EXEC|notify.AA_MAY_WRITE|notify.AA_MAY_READ|notify.AA_MAY_APPEND|notify.AA_MAY_CREATE) != 0 {
+		filePerms |= notify.AA_MAY_OPEN
+	}
+	return filePerms, nil
+}
+
+// ValidateConstraintsOutcomeLifespanExpiration returns an error if the given
+// constraints, outcome, lifespan, or duration are invalid, else returns nil.
+func ValidateConstraintsOutcomeLifespanExpiration(iface string, constraints *Constraints, outcome OutcomeType, lifespan LifespanType, expiration *time.Time, currTime time.Time) error {
+	if err := constraints.ValidateForInterface(iface); err != nil {
+		return err
+	}
+	if err := ValidateOutcome(outcome); err != nil {
+		return err
+	}
+	return ValidateLifespanExpiration(lifespan, expiration, currTime)
+}
+
+// ValidateConstraintsOutcomeLifespanDuration returns an error if the given
+// constraints, outcome, lifespan, or duration are invalid. Otherwise, converts
+// the given duration to an expiration timestamp and returns it and nil error.
+func ValidateConstraintsOutcomeLifespanDuration(iface string, constraints *Constraints, outcome OutcomeType, lifespan LifespanType, duration string) (*time.Time, error) {
```

Comment:
again it seems that this should be folded in the other taking a struct and make decision based on what lifespan is?

Reply:
This one is a bit more complex, as these fields are validated both when receiving a `PromptReply` from the client and when adding a new rule. I think the simplest/safest approach would be to continue to validate using `rule.Validate()` as part of `PopulateNewRule`, which is a helper in the requestrules called whenever a rule is being added or modified.

Rule expansion also occurs immediately after validation in requestrules, so `rule.Validate()` could also call `ExpandPathPattern`, making it all simpler.

The problem is that we want to be able to validate `PromptReply` contents before sending back a reply to the listener, all of which occurs before we go about creating a new rule. See: https://github.com/snapcore/snapd/pull/12964/files#diff-1bc7f26dd750993e7bf7bb0c8feb2e6d664f99313b5069f1dd68f4c3a66d51f9R234-R266

Reply:
So for now, I'm inclined to remove this function, and we can re-add something like it when it is known precisely what we need, likely during the upcoming `apparmorprompting` package PR.